src/Lecture1.hs

```diff
@@ -98,7 +103,17 @@ start position can be considered as zero (e.g. substring from the
 first character) and negative end position should result in an empty
 string.
 -}
-subString start end str = error "TODO"
+subString :: Int -> Int -> [Char] -> [Char]
+subString start end str
+    | end < 0 = []
+    | end < start = []
+    | otherwise = let x = getStart start in take (end-x+1) (drop x str)
+    where 
+        getStart :: Int -> Int
+        getStart st
+            | st < 0 = 0
+            | otherwise = st
```

Comment:
The `start` variable is visible in `where` so you can avoid passing an argument to `getStart`.
Moreover, you can use a standard function `max` to write `getStart` shorter.

Something like this:

```haskell
    | otherwise = take (end - actualStart + 1) (drop actualStart str)
    where 
        actualStart :: Int
        actualStart = max 0 start
```